README.md
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://github.com/Tencent/behaviac/blob/master/license.txt)
[![Release Version](https://img.shields.io/badge/release-3.6.39-red.svg)](https://github.com/Tencent/behaviac/releases)
[![Updates](https://img.shields.io/badge/Platform-%20iOS%20%7C%20OS%20X%20%7C%20Android%20%7C%20Windows%20%7C%20Linux%20-brightgreen.svg)](https://github.com/Tencent/behaviac/blob/master/history.txt)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/Tencent/behaviac/pulls)

## [访问http://www.behaviac.com/获取文档，教程，API，FAQ，源码，下载等一切相关资料](http://www.behaviac.com/)

 - behaviac是游戏AI的开发框架组件，也是游戏原型的快速设计工具
 - 支持行为树BT，状态机FSM，HTN等多种范式
 - 方便的编辑，实时和离线调试
 - 支持全平台，适用于客户端和服务器，助力游戏快速迭代开发
 - [官方网站](http://www.behaviac.com/)是文档，教程，API，FAQ，源码，下载等一切的入口
 - 您可以访问[behaviac Q&A](http://bbs.behaviac.com/)获取答案或提出问题
 - 您可以加入我们的QQ群433547396获得即时的帮助或者信息反馈。请大家优先使用[behaviac Q&A](http://bbs.behaviac.com/)。
 - BehaviacSetup*.exe是安装包，内含可执行的编辑器及示例。如需要自行构建，需要去[官方网站](http://www.behaviac.com/)或[github behaviac](https://github.com/Tencent/behaviac)下载或克隆源码，然后可以访问[build](http://www.behaviac.com/docs/zh/articles/build/)获取构建帮助
 

---------------------------------------------
## [Visist http://www.behaviac.com/ for documents, tutorials, FAQs and all other information](http://www.behaviac.com/)

 - behaviac is a framework of the game AI development, and it also can be used as a rapid game prototype design tool
 - behaviac supports the behavior tree, finite state machine and hierarchical task network
 - Behaviors can be designed and debugged in the designer, exported and executed by the game
 - The designer can only run on the Windows platforms, The run time library is implemented with C++ and C#, and it supports all major platforms (Windows, Linux, Android, iOS, Unity etc.) and Unity. 
 - The C++ version is suitable for the client and server side. 
 - [Website](http://www.behaviac.com/) for documents, tutorials, API，FAQ，source code, downloads，etc.
 - [behaviac Q&A](http://bbs.behaviac.com/) for answers of your questions.
 - BehaviacSetup*.exe is the setup package with the binary editor and demo executable. You can download/clone the source code from [github behaviac](https://github.com/Tencent/behaviac)
 
 


history.txt
2017-09-11  3.6.39
    Add the End node.
    Fix a bug for exporting the parameters of the method.
    Fix a bug for exporting the structs.

2017-08-17  3.6.38
    Improve the GC for the C#.

2017-08-08  3.6.37
    Support generating the codes for initializing the member array.

2017-08-03  3.6.36
    Fix a bug of accessing task params when calling a sub tree

2017-7-11  3.6.35
    Improve the C++ types generation.
    Fix a bug of time edition of the Wait node.
    Fix a bug of renaming the agent.
    Fix a bug of reentering the SelectorProbability node.
    Fix other bugs of the editor.

2017-5-16  3.6.34
    Fix a bug for the Parallel node.
    Fix a bug for exporting the return type of the method.
    Fix a bug for exporting the local vars.

2017-5-4  3.6.33
    Remove the inheritance for the agent type of the referenced behaviors.
    Fix a bug for the Parallel node.
    Fix a bug for the IfElse node.

2017-4-26  3.6.32
    Support resetting the children for the SelectorLoop node.
    Add the tutorial_13.

2017-4-24  3.6.31
    Fix a bug for generating the codes of the Task and Event.
    Fix a bug for the hotreload of the meta file.
    Add the tutorial_12.

2017-4-20  3.6.30
    Supporot the hot reload of the meta file.
    Ignore the System_Object as the type of the property.
    Fix a bug for the IfElse node.
    Fix a bug for generating the namespace of the enum.
    Add the tutorial_10 and tutorial_11.

2017-4-11  3.6.29
    Fix a bug of compiling error for GCC.

2017-4-10  3.6.28
    Improve generating the codes for the struct.
    Add tutorial_8 and tutorial_9.
    Fix a bug for editting the member of the struct.
    Fix a bug for creating an enum as a property.

2017-4-1  3.6.27
    Improve setting the properties of the node.
    Supporpt removing all instances when debugging.
    Fix a bug for filtering the properties when having selected one.

2017-3-23  3.6.26
    Do not send the update message when debugging.
    Fix a bug for setting the var when debugging for C#.
    Fix a bug for generating codes for Agent* as the return type of the method.
    Fix a bug for renaming the method.

2017-3-16  3.6.25
    Support renaming types.
    Add the tutorial_6 projects.

2017-3-10  3.6.24
    Add the tutorial_1_1 and tutorial_1_2.
    Add the Const setting for the parameters of the method.
    Improve the generated _Get_Property_ method.
    Fix a bug for the uint as the return type of the method.

2017-3-7  3.6.23
    Fix a bug for adding the struct of UnityEngine.Vector3.
    Fix a bug for the size_t value.
    Fix a bug for setting the prefab.

2017-3-2  3.6.22
    Fix a bug for exporting the codes for the UnityEngine.Vector3.
    Fix a bug for the char* type.
    Fix a bug for renaming the parameters of the method.
    Support setting the default value for the enum members.
    Support multi-level namespaces for C++.
    Support copying and pasting the struct type as the parameters.

2017-2-27  3.6.21
    Fix a bug for adding and removing the properties of the struct.

2017-2-24  3.6.20
    Fix a bug for renaming the agent and setting its base class.

2017-2-23  3.6.19
    Fix a bug for the CMakeList.
    Add the vcxproj.filters files for the C++ projects.

2017-2-22  3.6.18
    Update the license and header information.

2017-2-16  3.6.17
    Fix a bug for the struct as the parameters of the method.
    Fix a bug for setting the name for the property and method.
    Fix a bug for creating a new tree and setting its name.

2017-2-13  3.6.16
    Fix a bug for exporting the codes for getting the instance.

2017-2-8  3.6.15
    Improve the meta browser.
    Save the flag for checking error.
    Add the workspaces for the tutorials.

2017-2-5  3.6.14
    Fix a bug for the std::string and wstring.
    Fix a bug for the relative header path for the generated C++ codes.
    Preload for the referenced true and task.
    Remove all source code files from the install package.

2017-1-10  3.6.13
    Add tutorial_0/unity project.
    Improve creating and using the property.
	
2016-12-21  3.6.12
    Support the multiply meta files.
    Improve the codes for the TScan.
    Do not create the Prefabs folder if empty.

2016-12-16  3.6.11
    Add the IsRef for the type.
    Fix a bug for the child struct.
    Improve the codes for the Coverity.

2016-12-14  3.6.10
    Add MetaFile method for the Workspace.
    Fix a bug for loading the properties and methods.
    Rename "Prefabs".
    Support 14 parameters at most for the method of C++.
    Refine extending the existing struct.

2016-12-06  3.6.9
    Fix a bug for the ref type of the parameter.
    Support 14 parameters at most for the method of C#.

2016-12-02  3.6.8
    fix bug of possible no agents list when connecting
    fix no variables value update bug when connecting
    fix possible crash when editing value in editor when connecting
    int and double time refined
    BEHAVIAC_NOT_USE_UNITY added
    BEHAVIAC_NOT_USE_MONOBEHAVIOUR added
    improve the meta editting
    fix various bugs

2016-11-30  3.6.7
    refine mete handling

2016-11-28  3.6.6
    fix a bug when compare agent with null
    add more begin/end comment block for code generation
    agent/enum/struct can be configured to generate at different path

2016-11-25  3.6.5
    Support multi meta files.

2016-11-24  3.6.4
    Fix a bug for merging the meta to bb.

2016-11-23  3.6.3
    Continue fixing bugs for types generation.

2016-11-21  3.6.2
    Fix bugs for types generation.

2016-11-18  3.6.1
    Improve the Agent types generation.
    Refactor the C++ runtime.
    Fix the GC problem for the enum and struct comparation.

2016-11-10  3.6.0
    Improve the Agent types generation.
    Refactor the C++ runtime.
    Fix the GC problem for the enum and struct comparation.
    adjust files structure for the next release

2016-11-3  3.5.18
    Fix a bug for replacing the node on the Prefab.
    Do not export the SetVariable() for the assignment of the property.

src/common/crc.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/common/base.h"
#include "behaviac/common/assert.h"

#include "behaviac/common/crc.h"

#include <ctype.h>

namespace behaviac {
    namespace CRC32 {
        // pre genereated by make_hash_table
        static uint32_t s_crcTable[] = {
            0u, 1996959894u, 3993919788u, 2567524794u, 124634137u, 1886057615u, 3915621685u, 2657392035u, 249268274u, 2044508324u, 3772115230u, 2547177864u, 162941995u, 2125561021u, 3887607047u, 2428444049u,
            498536548u, 1789927666u, 4089016648u, 2227061214u, 450548861u, 1843258603u, 4107580753u, 2211677639u, 325883990u, 1684777152u, 4251122042u, 2321926636u, 335633487u, 1661365465u, 4195302755u, 2366115317u,
            997073096u, 1281953886u, 3579855332u, 2724688242u, 1006888145u, 1258607687u, 3524101629u, 2768942443u, 901097722u, 1119000684u, 3686517206u, 2898065728u, 853044451u, 1172266101u, 3705015759u, 2882616665u,
            651767980u, 1373503546u, 3369554304u, 3218104598u, 565507253u, 1454621731u, 3485111705u, 3099436303u, 671266974u, 1594198024u, 3322730930u, 2970347812u, 795835527u, 1483230225u, 3244367275u, 3060149565u,
            1994146192u, 31158534u, 2563907772u, 4023717930u, 1907459465u, 112637215u, 2680153253u, 3904427059u, 2013776290u, 251722036u, 2517215374u, 3775830040u, 2137656763u, 141376813u, 2439277719u, 3865271297u,
            1802195444u, 476864866u, 2238001368u, 4066508878u, 1812370925u, 453092731u, 2181625025u, 4111451223u, 1706088902u, 314042704u, 2344532202u, 4240017532u, 1658658271u, 366619977u, 2362670323u, 4224994405u,
            1303535960u, 984961486u, 2747007092u, 3569037538u, 1256170817u, 1037604311u, 2765210733u, 3554079995u, 1131014506u, 879679996u, 2909243462u, 3663771856u, 1141124467u, 855842277u, 2852801631u, 3708648649u,
            1342533948u, 654459306u, 3188396048u, 3373015174u, 1466479909u, 544179635u, 3110523913u, 3462522015u, 1591671054u, 702138776u, 2966460450u, 3352799412u, 1504918807u, 783551873u, 3082640443u, 3233442989u,
            3988292384u, 2596254646u, 62317068u, 1957810842u, 3939845945u, 2647816111u, 81470997u, 1943803523u, 3814918930u, 2489596804u, 225274430u, 2053790376u, 3826175755u, 2466906013u, 167816743u, 2097651377u,
            4027552580u, 2265490386u, 503444072u, 1762050814u, 4150417245u, 2154129355u, 426522225u, 1852507879u, 4275313526u, 2312317920u, 282753626u, 1742555852u, 4189708143u, 2394877945u, 397917763u, 1622183637u,
            3604390888u, 2714866558u, 953729732u, 1340076626u, 3518719985u, 2797360999u, 1068828381u, 1219638859u, 3624741850u, 2936675148u, 906185462u, 1090812512u, 3747672003u, 2825379669u, 829329135u, 1181335161u,
            3412177804u, 3160834842u, 628085408u, 1382605366u, 3423369109u, 3138078467u, 570562233u, 1426400815u, 3317316542u, 2998733608u, 733239954u, 1555261956u, 3268935591u, 3050360625u, 752459403u, 1541320221u,
            2607071920u, 3965973030u, 1969922972u, 40735498u, 2617837225u, 3943577151u, 1913087877u, 83908371u, 2512341634u, 3803740692u, 2075208622u, 213261112u, 2463272603u, 3855990285u, 2094854071u, 198958881u,
            2262029012u, 4057260610u, 1759359992u, 534414190u, 2176718541u, 4139329115u, 1873836001u, 414664567u, 2282248934u, 4279200368u, 1711684554u, 285281116u, 2405801727u, 4167216745u, 1634467795u, 376229701u,
            2685067896u, 3608007406u, 1308918612u, 956543938u, 2808555105u, 3495958263u, 1231636301u, 1047427035u, 2932959818u, 3654703836u, 1088359270u, 936918000u, 2847714899u, 3736837829u, 1202900863u, 817233897u,
            3183342108u, 3401237130u, 1404277552u, 615818150u, 3134207493u, 3453421203u, 1423857449u, 601450431u, 3009837614u, 3294710456u, 1567103746u, 711928724u, 3020668471u, 3272380065u, 1510334235u, 755167117u
        };

        //void init_crc_table(void)
        //{
        //	// Standard CRC-32 ppolynomial
        //	const uint32_t POLYNOMIAL = 0x04c11db7;

        //	uint32_t crc_accum = 0;

        //	for (int i = 0; i < 256; i++)
        //	{
        //		crc_accum = (i << 24);
        //		for (int j = 0; j < 8; j++)
        //		{
        //			if (crc_accum & 0x80000000L)
        //				crc_accum = (crc_accum << 1) ^ POLYNOMIAL;
        //			else
        //				crc_accum = (crc_accum << 1);
        //		}

        //		s_crcTable[i] = crc_accum;
        //	}
        //}

        ////Generate a table for a byte-wise 32-bit CRC calculation on the polynomial:
        ////x^32+x^26+x^23+x^22+x^16+x^12+x^11+x^10+x^8+x^7+x^5+x^4+x^2+x+1.
        //void make_hash_table()
        //{
        //	// terms of polynomial defining this crc (except x^32):
        //	const char polynomial_coef[] = { 0, 1, 2, 4, 5, 7, 8, 10, 11, 12, 16, 22, 23, 26 };

        //	//make exclusive-or pattern from polynomial (0xedb88320L)
        //	unsigned long poly = 0L;
        //	for (int n = 0; n < sizeof(polynomial_coef) / sizeof(char); n++) {
        //		poly |= 1L << (31 - polynomial_coef[n]);
        //	}

        //	for (int n = 0; n < 256; n++) {
        //		unsigned long c = (unsigned long)n;
        //		for (int k = 0; k < 8; k++) {
        //			c = c & 1 ? poly ^ (c >> 1) : c >> 1;
        //		}

        //		s_crcTable[n] = c;
        //	}
        //}

        //unsigned int crc_string(unsigned char *message) {
        //	int i = 0;
        //	unsigned int crc_result = 0xFFFFFFFF;
        //	while (message[i] != 0) {
        //		unsigned int byte = message[i];
        //		crc_result = crc_result ^ byte;
        //		for (int j = 7; j >= 0; j--) {
        //			unsigned int mask = (unsigned int)(-(int)(crc_result & 1));
        //			crc_result = (crc_result >> 1) ^ (0xEDB88320 & mask);
        //		}

        //		i = i + 1;
        //	}

        //	return ~crc_result;
        //}

        static void UpdateB(uint32_t&  crc, uint8_t b) {
            crc = s_crcTable[(crc ^ b) & 0XFF] ^ (crc >> 8);
        }

        //static void Update(uint32_t& crc, uint32_t d)
        //{
        //    UpdateB(crc, (uint8_t)(d & 0XFF));
        //    UpdateB(crc, (uint8_t)((d >> 8) & 0XFF));
        //    UpdateB(crc, (uint8_t)((d >> 16) & 0XFF));
        //    UpdateB(crc, (uint8_t)((d >> 24) & 0XFF));
        //}

        //static void Finish(uint32_t& crc)
        //{
        //    crc ^= 0XFFFFFFFF;
        //}

        uint32_t CalcCRC(const char* str) {
            uint32_t crc = 0XFFFFFFFF;
            size_t length = strlen(str);

            for (size_t i = 0; i < length; ++i) {
                char ch = (uint8_t)str[i];
                uint8_t b = (uint8_t)ch;

                UpdateB(crc, b);
            }

            UpdateB(crc, (uint8_t)length);
            return crc;
        }

        //uint32_t CalcCRCNoCase(const char* str)
        //{
        //	uint32_t crc = 0XFFFFFFFF;
        //	size_t length = strlen(str);

        //	for (size_t i = 0; i < length; ++i)
        //	{
        //		char ch = (uint8_t)str[i];
        //		uint8_t b = (uint8_t)tolower(ch);

        //		UpdateB(crc, b);
        //	}

        //	UpdateB(crc, (uint8_t)length);
        //	return crc;
        //}
    }//namespace CRC32
}//
